File: hashtable/438.go
func findAnagrams(s string, p string) []int {
    target := make(map[byte]int)
    ret := make([]int, 0)
    var counter int
    for i := 0; i < len(p); i++ {
        if _, ok := target[p[i]]; ok {
            target[p[i]]++
        }else {
            target[p[i]] = 1
            counter++
        }
    }
    
    var start, end int
    for end < len(s) {
        if _, ok := target[s[end]]; ok {
            target[s[end]]--
            if target[s[end]] == 0 {
                counter--
            }
        }
        end++
        
        for counter == 0 {
            if _, ok := target[s[start]]; ok {
                target[s[start]]++
                if target[s[start]] > 0 {
                    counter++
                }
            }
            if end - start == len(p) {
                ret = append(ret, start)
            }
            start++
        }
    }
    return ret
}

File: hashtable/438.cc
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        std::unordered_map<char, int> m;
        int counter = 0;
        for (const auto& c : p) {
            auto iter = m.find(c);
            if (iter == m.end()) {
                ++counter;
                m[c] = 1;
            } else {
                ++m[c];
            }
        }
        int begin = 0, end = 0;
        std::vector<int> ret;
        while (end < s.size()) {
            auto iter_end = m.find(s[end]);
            if (iter_end != m.end()) {
                --m[s[end]];
                if (0 == m[s[end]]) --counter;
            }
            ++end;
            while (0 == counter) {
                auto iter_begin = m.find(s[begin]);
                if (iter_begin != m.end()) {
                    ++m[s[begin]];
                    if (m[s[begin]] > 0) ++counter;
                }
                if ((end - begin) == p.size()) 
                    ret.emplace_back(begin);
                ++begin;
            }
        }
        return  ret;
    }
};
